src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -173,15 +179,60 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
                         if (InitializationCacheExpires < DateTimeOffset.UtcNow)
                         {
                             InitializationCacheExpires = DateTimeOffset.UtcNow.Add(MinCacheExpirationInterval);
-                            await LoadAll(ignoreFailures: false, cancellationToken).ConfigureAwait(false);
+                            await LoadAllWithFailOverPolicyAsync(ignoreFailures: false, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        return;
+                    }
+
+                    async Task<(bool, List<ConfigurationSetting>, Dictionary<KeyValueWatcher, KeyValueChange>, List<KeyValueChange>)> funcToExecute(ConfigurationClient client)
+                    {
+                        (bool shouldRefreshAll,
+                         List<ConfigurationSetting> changedAdapterSettings,
+                         Dictionary<KeyValueWatcher, KeyValueChange> changedKeyValues) = await GetRefreshedIndividualKeyValues(client, cancellationToken).ConfigureAwait(false);
+
+                        if (shouldRefreshAll)
+                        {
+                            return (shouldRefreshAll, changedAdapterSettings, changedKeyValues, null);
                         }
 
+                        List<KeyValueChange> changedKeyValuesCollection = await GetRefreshedKeyValueCollections(client, cancellationToken).ConfigureAwait(false);
+
+                        return (shouldRefreshAll, changedAdapterSettings, changedKeyValues, changedKeyValuesCollection);
+                    }
+
+                    (bool shouldRefreshAll, List<ConfigurationSetting> changedAdapterSettings, Dictionary<KeyValueWatcher, KeyValueChange> changedKeyValues, List<KeyValueChange> changedKeyValueCollections) = await ExecuteWithFailOverPolicyAsync(funcToExecute, cancellationToken).ConfigureAwait(false);
+
+                    // Trigger a single refresh-all operation if a change was detected in one or more key-values with refreshAll: true
+                    if (shouldRefreshAll)
+                    {
+                        await LoadAllWithFailOverPolicyAsync(ignoreFailures: false, cancellationToken).ConfigureAwait(false);

```

Comment:
I was thinking that since we are updating everything, we may as well utilize all clients to update all clients, but the issue you brought up makes sense. Updated the logic now.